src/loop_rerolling.py

```diff
@@ -0,0 +1,116 @@
+from typing import List
+
+from .flow_graph import (
+    BasicNode,
+    ConditionalNode,
+    FlowGraph,
+    Node,
+    compute_relations,
+)
+from .parse_instruction import Instruction
+
+
+def replace_node_references(
+    flow_graph: FlowGraph, replace_this: Node, with_this: Node
+) -> None:
+    for node_to_modify in flow_graph.nodes:
+        node_to_modify.replace_any_children(replace_this, with_this)
+
+
+def remove_node(flow_graph: FlowGraph, to_delete: Node, new_child: Node) -> None:
+    flow_graph.nodes.remove(to_delete)
+    replace_node_references(flow_graph, to_delete, new_child)
+
+
+def replace_node(flow_graph: FlowGraph, replace_this: Node, with_this: Node) -> None:
+    replacement_index = flow_graph.nodes.index(replace_this)
+    flow_graph.nodes[replacement_index] = with_this
+    replace_node_references(flow_graph, replace_this, with_this)
+
+
+def reroll_loop(flow_graph: FlowGraph, start: ConditionalNode) -> bool:
```

Comment:
Add a comment with the structure/asm we're trying to pattern-match here (or refer to a test?).

Reply:
It'd be nice to include a `dot` representation of what's going on:

```dot
    digraph before {
        start -> node_1 [label=F, color=blue];
        start -> node_7 [label=T, color=red];
        node_1 -> node_2 [label=F, color=blue];
        node_1 -> node_5 [label=T, color=red];
        node_2 -> node_3;
        node_3 -> node_4 [label=F, color=blue];
        node_3 -> node_3 [label=T, color=red];
        node_4 -> node_5 [label=F, color=blue];
        node_4 -> node_7 [label=T, color=red];
        node_5 -> node_6;
        node_6 -> node_7 [label=F, color=blue];
        node_6 -> node_6 [label=T, color=red];
    }

    digraph after {
        start -> new_node_1 [label=F, color=blue];
        start -> node_7 [label=T, color=red];
        new_node_1 -> node_2 [label=F, color=blue];
        new_node_1 -> node_7 [label=T, color=red];
        node_2 -> node_3;
        node_3 -> node_7 [label=F, color=blue];
        node_3 -> node_3 [label=T, color=red];
    }
```
And maybe a link to an online render, like https://bit.ly/3A2MEJt

Reply:
This is actually nearly exactly what the new reroll.py does -- stay tuned